From a49ecd72568dae45b40b8771fdcf416f2e2ad4d7 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Tue, 24 Jun 2025 13:08:07 +0800
Subject: [PATCH 26/63] loonggpu: Handle tt moves properly

It seems required since a TTM move refactoring in 2020.  Following up the
reference implementation change (and several changes after that).

Link: https://git.kernel.org/torvalds/c/3a08446b31e3
Link: https://git.kernel.org/torvalds/c/c37d951cb42a
Link: https://git.kernel.org/torvalds/c/9764c35348b4
Link: https://git.kernel.org/torvalds/c/29a1d482e404
Link: https://git.kernel.org/torvalds/c/18ab7e88778f
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 loonggpu/loonggpu_ttm.c | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/loonggpu/loonggpu_ttm.c b/loonggpu/loonggpu_ttm.c
index 783ba2a..6bc7d99 100644
--- a/loonggpu/loonggpu_ttm.c
+++ b/loonggpu/loonggpu_ttm.c
@@ -610,6 +610,8 @@ out_cleanup:
 	return r;
 }
 
+static lg_ttm_backend_unbind_ret loonggpu_ttm_backend_unbind(lg_ttm_backend_func_arg);
+
 /**
  * loonggpu_bo_move - Move a buffer object to a new memory location
  *
@@ -643,15 +645,24 @@ static int loonggpu_bo_move(struct ttm_buffer_object *bo, bool evict,
 		goto out;
 	}
 
-	if ((old_mem->mem_type == TTM_PL_TT &&
-	     new_mem->mem_type == TTM_PL_SYSTEM) ||
-	    (old_mem->mem_type == TTM_PL_SYSTEM &&
-	     new_mem->mem_type == TTM_PL_TT)) {
+	if (old_mem->mem_type == TTM_PL_SYSTEM &&
+	    new_mem->mem_type == TTM_PL_TT) {
 		/* bind is enough */
 		loonggpu_move_null(bo, new_mem);
 		goto out;
 	}
 
+	if (old_mem->mem_type == TTM_PL_TT &&
+	    new_mem->mem_type == TTM_PL_SYSTEM) {
+		r = ttm_bo_wait_ctx(bo, ctx);
+		if (r)
+			return r;
+
+		loonggpu_ttm_backend_unbind(bo->bdev, bo->ttm);
+		loonggpu_move_null(bo, new_mem);
+		goto out; // XXX this is not in LG110, maybe a mistake
+	}
+
 	if (old_mem->mem_type == LOONGGPU_PL_DOORBELL ||
 	    new_mem->mem_type == LOONGGPU_PL_DOORBELL) {
 		/* Nothing to save here */
-- 
2.52.0

